src/rand/rand.zig

```diff
@@ -114,6 +114,66 @@ pub fn BlockRng(
     };
 }
 
+/// Downsample a random number generator to a smaller range.
+/// This implementationc is based on the implementation in the rust rand crate
+/// and ensures the same sequence is generated.
+pub fn uintLessThanRust(r: Random, comptime T: type, less_than: T) T {
+    comptime std.debug.assert(@typeInfo(T).Int.signedness == .unsigned);
+    const bits = @typeInfo(T).Int.bits;
+    const max = std.math.maxInt(T);
+    std.debug.assert(0 < less_than);
+
+    const z = (max - less_than + 1) % less_than;
+    const zone = max - z;
+    while (true) {
+        const v = r.int(T);
+        const m = std.math.mulWide(T, v, less_than);
+        const lo: T = @truncate(m);
+        if (lo <= zone) {
+            return @truncate(m >> bits);
+        }
+    }
+
+    unreachable;
```

Comment:
I don't think this `unreachable` is necessary, the compiler can already prove that a `break`less `while (const_condition)` never reaches this.

Reply:
I got a compile error without it, thats why I chucked it in their, its not very nice though I agree

Reply:
[060375](https://github.com/Syndica/sig/pull/211/commits/060375c68e4b0d08b9f6b52c6a2047f9ae00765c)